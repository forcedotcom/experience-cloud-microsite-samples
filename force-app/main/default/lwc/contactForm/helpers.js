/**
 * Utility to handle redirect action
 *
 * @param {string} url Redirect url
 */
export function redirectPage(url) {
    window.open(url, '_self');
}

/**
 * Utility to verify form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Boolean} To show whether all inputs are valid
 */
export function isSubmitActionDataValid(inputs) {
    return inputs.every((input) => input?.validity?.valid);
}

/**
 * Utility to get form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Object} Object mapping fields to values
 */
export function transformFormData(inputs) {
    const fields = {};
    inputs.forEach((element) => {
        fields[element.name] = element.value;
    });
    return fields;
}

/**
 * Utility to format form data for contact form submission
 *
 * @param {object} fieldsData This is an object representing form fields. Eg. {FirstName: John, LastName: Doe}
 * @returns {object} Data formatted correctly and ready for insertion to marketing cloud
 */
export function transformFieldsForContactFormSubmission(fieldsData) {
    const formFields = Object.keys(fieldsData ?? {}).map((fieldName) => {
        const fieldValue = normalizeValue(fieldsData[fieldName]);
        return {
            name: fieldName,
            value: fieldValue
        };
    });
    return {
        formFieldsList: {
            formFields
        }
    };
}

/**
 * Utility to normalize all field values to text (all contact form fields are currently Text or Email)
 *
 * @param {string|boolean} value The form field to normalize
 * @returns {string} The string representation of the field
 */
export function normalizeValue(value) {
    if (value === null || value === undefined) {
        return '';
    }
    return String(value);
}

/**
 * Utility to check validity of contact form data
 *
 * @param {object} formInput The input for the contact Form
 * @returns {boolean} Is the form Input Data valid ?
 */
export function validateFormData(formInput) {
    const formFields = formInput?.formFieldsList?.formFields;
    return Array.isArray(formFields);
}
