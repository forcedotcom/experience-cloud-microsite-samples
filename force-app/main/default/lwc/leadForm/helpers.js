/**
 * Utility to handle redirect action
 *
 * @param {string} url Redirect url
 */
export function redirectPage(url) {
    window.open(url, '_self');
}

/**
 * Utility to verify form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Boolean} To show whether all inputs are valid
 */
export function isSubmitActionDataValid(inputs) {
    const invalids = inputs?.filter((element) => !element.validity.valid);

    if (invalids?.length > 0) {
        invalids.forEach((element) => element.reportValidity());
        return false;
    }
    return true;
}

/**
 * Utility to get form data
 *
 * @param {Array} inputs Lightning inputs from the form
 * @returns {Object} Object mapping Lead fields with values
 */
export function getFormData(inputs) {
    const fields = {};
    inputs.forEach((element) => {
        fields[element.name] =
            element.type === 'checkbox' ? element.checked : element.value;
    });
    return fields;
}
