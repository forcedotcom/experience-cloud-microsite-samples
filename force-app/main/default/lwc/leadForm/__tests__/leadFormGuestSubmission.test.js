import { createElement } from 'lwc';
import LeadForm from 'c/leadForm';
import { createRecord } from 'lightning/uiRecordApi';
import * as helper from '../helpers';
import { DEFAULT_LEAD_FORM_PROPERTIES } from './testUtils';

/**
 * @param {object} props object with all properties for lead form component
 * @returns {object} - HTML DOM object
 */
function createComponent(props = {}) {
    const element = createElement('c-lead-form', {
        is: LeadForm
    });
    Object.assign(element, props);
    document.body.appendChild(element);
    return element;
}

/**
 * Helper sets required fields with default values
 *
 * @param {object} element - HTML DOM object reference
 */
function setDefaultValues(element) {
    //Set company default value
    const companyInput = element.shadowRoot.querySelector(
        'lightning-input[data-company]'
    );
    companyInput.value = 'Salesforce';

    //Set lastname default value
    const lastnameInput = element.shadowRoot.querySelector(
        'lightning-input[data-lastname]'
    );
    lastnameInput.value = 'Smith';
}

describe('Lead Form Component Guest User Submit Tests', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
        // Prevent data saved on mocks from leaking between tests
        jest.clearAllMocks();
    });

    it('should create a lead successfully with a 404 error when logged in as a guest user ', async () => {
        //Create component and set required fields with default values
        const element = createComponent(DEFAULT_LEAD_FORM_PROPERTIES);
        setDefaultValues(element);

        //Mocking isSubmitActionDataValid
        const spy = jest.spyOn(helper, 'isSubmitActionDataValid');
        spy.mockImplementation(() => {
            return true;
        });

        //Mock createRecord throwing an error
        const err = new Error('Error!');
        err.status = 404;
        createRecord.mockRejectedValue(err);

        //Initiate submit logic
        const form = element.shadowRoot.querySelector('form');
        form.dispatchEvent(new CustomEvent('submit'));

        //Resolve promise for API call
        await Promise.resolve();
        //Need to wait for then on API call
        await Promise.resolve();
        //Need to wait for then on the API callout logic
        await Promise.resolve();
        //Need to wait for flags to be changed in order to update UI
        await Promise.resolve();
        //Need to wait for lwc re-rendering
        await Promise.resolve();

        expect(element).toMatchSnapshot();

        //Check if success message is present
        const pTag = element.shadowRoot.querySelector('p');
        expect(pTag.textContent).toEqual(
            DEFAULT_LEAD_FORM_PROPERTIES.submitActionSuccessMessage
        );
    });
});
