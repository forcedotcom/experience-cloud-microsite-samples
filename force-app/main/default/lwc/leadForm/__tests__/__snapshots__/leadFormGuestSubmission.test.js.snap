// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lead Form Component Guest User Submit Tests should create a lead successfully with a 404 error when logged in as a guest user  1`] = `
<c-lead-form>
  #shadow-root(open)
    <p
      class="slds-text-color_success slds-var-m-left_medium"
      role="status"
    >
      Thank you!
    </p>
</c-lead-form>
`;
